pkg/services/opsgenie/opsgenie_test.go

```diff
@@ -0,0 +1,417 @@
+						`"tags":["tag1","tag2"],` +
+						`"details":{"key1":"value1","key2":"value2"},` +
+						`"entity":"1",` +
+						`"source":"1",` +
+						`"priority":"P1",` +
+						`"user":"1",` +
+						`"note":"1"` +
+						`}`))
+				}
+
+				err := service.Send("1", &types.Params{
+					"alias":       "1",
+					"description": "1",
+					"responders":  "team:1",
+					"visibleTo":   "team:1",
+					"actions":     "action1,action2",
+					"tags":        "tag1,tag2",
+					"details":     "key1:value1,key2:value2",
+					"entity":      "1",
+					"source":      "1",
+					"priority":    "P1",
+					"user":        "1",
+					"note":        "1",
+				})
+				Expect(err).To(BeNil())
+
+				checkRequest = func(body string, header http.Header) {
+					Expect(header["Authorization"][0]).To(Equal("GenieKey " + mockAPIKey))
+					Expect(header["Content-Type"][0]).To(Equal("application/json"))
+					Expect(body).To(Equal(`{` +
+						`"message":"2",` +
+						`"alias":"query-alias",` +
+						`"description":"query-description",` +
+						`"responders":[{"type":"team","name":"query_team"}],` +
+						`"visibleTo":[{"type":"user","username":"query_user"}],` +
+						`"actions":["queryAction1","queryAction2"],` +
+						`"tags":["queryTag1","queryTag2"],` +
+						`"details":{"queryKey1":"queryValue1","queryKey2":"queryValue2"},` +
+						`"entity":"query-entity",` +
+						`"source":"query-source",` +
+						`"priority":"P2",` +
+						`"user":"query-user",` +
+						`"note":"query-note"` +
+						`}`))
+				}
+
+				err = service.Send("2", nil)
+				Expect(err).To(BeNil())
+			})
+		})
+	})
+})
+
+var _ = Describe("the OpsGenie Config struct", func() {
+	When("generating a config from a simple URL", func() {
+		It("should populate the config with host and apikey", func() {
+			url, err := url.Parse(fmt.Sprintf("opsgenie://%s/%s", mockHost, mockAPIKey))
+			Expect(err).To(BeNil())
+
+			config := Config{}
+			err = config.SetURL(url)
+			Expect(err).To(BeNil())
+
+			Expect(config.ApiKey).To(Equal(mockAPIKey))
+			Expect(config.Host).To(Equal(mockHost))
+			Expect(config.Port).To(Equal(uint16(0)))
+		})
+	})
+
+	When("generating a config from a url with port", func() {
+		It("should populate the port field", func() {
+			url, err := url.Parse(fmt.Sprintf("opsgenie://%s:12345/%s", mockHost, mockAPIKey))
+			Expect(err).To(BeNil())
+
+			config := Config{}
+			err = config.SetURL(url)
+			Expect(err).To(BeNil())
+
+			Expect(config.Port).To(Equal(uint16(12345)))
+		})
+	})
+
+	When("generating a config from a url with query parameters", func() {
+		It("should populate the config fields with the query parameter values", func() {
+			queryParams := `alias=Life+is+too+short+for+no+alias&description=Every+alert+needs+a+description&actions=An+action&tags=tag1,tag2&details=key:value,key2:value2&entity=An+example+entity&source=The+source&priority=P1&user=Dracula&note=Here+is+a+note&responders=user:Test,team:NOC&visibleTo=user:A+User`
```

Comment:
This would make the test work with the encoded query field values:
```suggestion
			queryParams := `alias=Life+is+too+short+for+no+alias&description=Every+alert+needs+a+description&actions=An+action&tags=tag1%2Ctag2&details=key%3Avalue%2Ckey2%3Avalue2&entity=An+example+entity&source=The+source&priority=P1&user=Dracula&note=Here+is+a+note&responders=user%3ATest%2Cteam%3ANOC&visibleTo=user%3AA+User`
```
The users won't have to escape all the characters, it will handle both, but the generated ones are at least ensured to always be valid URL query params.

Reply:
Ok fixed. I misunderstood and thought you wanted to preserve the commas, etc. when generating urls.

Reply:
> If you move out the looping from the entity methods to a pair of

Done, is that what you had in mind?

Reply:
> If you move out the looping from the entity methods to a pair of:
That was one mess of a sentence. But yeah, exactly.